pkg/boot/stboot/bootball.go

```diff
@@ -0,0 +1,574 @@
+	}
+	return
+}
+
+func getConfig(dest string) (cfg *Stconfig, err error) {
+	cfgBytes, err := ioutil.ReadFile(dest)
+	if err != nil {
+		return
+	}
+	cfg, err = stconfigFromBytes(cfgBytes)
+	if err != nil {
+		return
+	}
+	if !(cfg.IsValid()) {
+		return cfg, errors.New("invalid configuration")
+	}
+	return
+}
+
+func getRootCert(dest string) (cert *x509.CertPool, err error) {
+	certBytes, err := ioutil.ReadFile(dest)
+	if err != nil {
+		return
+	}
+	cert, err = certPool(certBytes)
+	if err != nil {
+		return
+	}
+	return
+}
+
+func getBootFiles(cfg *Stconfig, prefix string) (bootFiles map[string][]string, err error) {
+	bootFiles = make(map[string][]string)
+	for _, bc := range cfg.BootConfigs {
+		files := make([]string, 0)
+		for _, file := range bc.FileNames() {
+			file = filepath.Join(prefix, file)
+			files = append(files, file)
+			if err = validateFiles("", files...); err != nil {
+				return
+			}
+		}
+		bootFiles[bc.Name] = files
+	}
+	return
+}
+
+func (ball *BootBall) getSignatures() (err error) {
+	ball.signatures = make(map[string][]signature)
+	path := filepath.Join(ball.dir, signaturesDirName)
+
+	sigPool := make([]signature, 0)
+	err = filepath.Walk(path, func(path string, info os.FileInfo, err error) error {
+		ext := filepath.Ext(info.Name())
+
+		if !info.IsDir() && (ext == ".signature") {
+			sigBytes, err := ioutil.ReadFile(path)
+			if err != nil {
+				return err
+			}
+
+			dir := filepath.Dir(path)
+			index, err := strconv.Atoi(dir[len(dir)-1:])
+			if err != nil {
+				return err
+			}
+
+			certFile := strings.TrimSuffix(path, filepath.Ext(path)) + ".cert"
+			certBytes, err := ioutil.ReadFile(certFile)
+			if err != nil {
+				return err
+			}
+
+			cert, err := parseCertificate(certBytes)
+			if err != nil {
+				return err
+			}
+
+			sig := signature{
+				Bytes: sigBytes,
+				Cert:  cert,
+			}
+			sigPool = append(sigPool, sig)
+			key := ball.config.BootConfigs[index].Name
+			ball.signatures[key] = sigPool
+		}
+		return nil
+	})
+	if err != nil {
+		return
+	}
+	return
+}
+
+func validateFiles(prefix string, files ...string) (err error) {
+	for _, file := range files {
+		_, err = os.Stat(filepath.Join(prefix, file))
+		if err != nil {
+			return
+		}
+	}
+	return
+}
+
+func writeSignatures(sigs []signature, certFile, dir string) (err error) {
+	for i, sig := range sigs {
+		d := fmt.Sprintf("%s%d", bootFilesDirName, i)
+		path := filepath.Join(dir, signaturesDirName, d)
+		os.Mkdir(path, os.ModePerm)
+
+		id := fmt.Sprintf("%x", sig.Cert.PublicKey)[2:18]
+		sigName := fmt.Sprintf("%s.signature", id)
+		sigPath := filepath.Join(path, sigName)
+		werr := ioutil.WriteFile(sigPath, sig.Bytes, 0644)
+		if werr != nil {
+			return werr
+		}
+
+		certName := fmt.Sprintf("%s.cert", id)
+		certPath := filepath.Join(path, certName)
+		cerr := copyFile(certFile, certPath)
+		if cerr != nil {
+			return cerr
+		}
+	}
+	return
+}
+
+func makeConfigDir(cfg *Stconfig, origDir string) (dir string, err error) {
```

Comment:
makeConfigDir copies the files named in cfg to well known directory tree inside the bootball's underlying tmpDir since the files in user's cfg can reside anywhere in the file system.
It is used when initializing a BootBall from a user's stconfig.json file and prepares the BootBall to be packed to an archive in a standardized manner.